### ISSUE TITLE
[Request] USB trigger

### ISSUE THREAD

**Issue by InvisibleMan101** [x13a/Wasted#6]
Thank you for all your work & dedication. Is it possible to implement:

- Duress Password Wipe
- USB extraction wipe

Not sure if these are something that might be in the works  moving forward but thanks again either way.

**Comment by x13a** [comment]
You have to explain logic by "Duress Password Wipe".

**Comment by InvisibleMan101** [comment]
> You have to explain logic by "Duress Password Wipe".

My apologies, enter a certain password on the lock screen and have device wipe itself or initiate factory reset.

**Comment by x13a** [comment]
I suppose it can not be done. Maybe something will change in Android 13. But USB trigger is likely to be implemented)

**Comment by InvisibleMan101** [comment]
> I suppose it can not be done. Maybe something will change in Android 13. But USB trigger is likely to be implemented)

That's okay I just assumed that maybe if you needed to wipe your device or give a political adversary your password then this function would be of some use but perhaps maybe not as it may get you in more hot water if it ever came to that point and if a password that when entered wiped your device then perhaps that feature may not serve much purpose if everything else is working that this project provides and the device is encrypted.

What you have made for now is more than enough. Thank you again for your work, looking forward to USB trigger :)

**Comment by x13a** [comment]
It is an Android limitation, not my. I have done something similar for linux in the past. You can set maximum failed password attempts to 1 and be very carefull.

**Comment by InvisibleMan101** [comment]
> It is an Android limitation, not my. I have done something similar for linux in the past. You can set maximum failed password attempts to 1 and be very carefull.

Thank you that makes perfect sense. Yes like in Linux similar with the nuke password wipe option. But your right hopefully something changes in Android 13 to remove the limitation.

**Comment by gostaex** [comment]
Duress Password Wipe
This is already implemented in this program. The password is called negation and it works. Maybe just add some code. 
https://github.com/nekohasekai/lockup

**Comment by gostaex** [comment]
Please tell me how to use the notification label and tile functions

**Comment by x13a** [comment]
> This is already implemented in this program.

I will take a look, thanks.

> how to use the notification label and tile functions

After you enable `Notification` in `Wasted`, you have to grant `read notifications` permission. Then you can send a message with authentication code and boom) You can test this with unchecked `Wipe data` option.

Tile is useful when you device is locked, cause it can be used from lock screen. After you click on the tile (it is an Android buttons you get when swipe down from top of the screen, with functions like disable Wi-Fi and so on) this tile become active and where are delay for 2 seconds. If you click again this will cancel trigger, else run lock/wipe routine.

**Comment by gostaex** [comment]
Thank you understood. and there is also a shortcut button. He should appear.

**Comment by x13a** [comment]
Yes, make a long click on the app's icon and you will see a `Panic` shortcut.

**Comment by gostaex** [comment]
Yes, everything worked out. label works. But with the tile, nothing happened. There is no label on the plate. Or you just need to swipe a few times

**Comment by x13a** [comment]
![1](https://user-images.githubusercontent.com/53379023/151036850-dc5b10ed-38a1-4483-b4fa-15812f503dd6.png)

**Comment by gostaex** [comment]
Yes thank you. Understood.
Thank you very much for your work.

**Comment by x13a** [comment]
@gostaex , in the next release Tile will mimic to "Airplane mode" be carefull.

**Comment by gostaex** [comment]
164 / 5 000
Результаты перевода
Fine. Yes, this is an interesting idea. And what about injecting a function from lockup when a password reset occurs. There the author uploads his code under a free license.

**Comment by x13a** [comment]
Now implementing USB trigger. After it will see.

**Comment by InvisibleMan101** [comment]
> Fine. Yes, this is an interesting idea. And what about injecting a function from lockup when a password reset occurs. There the author uploads his code under a free license.

 I did notice actually in the latest android release of LOS on one of my test devices that lockup does not work. I think this was version 12 and possibly onward even when you enter a duress password it just doesn't do anything at all but on android 11 and below no issues and it works as it should.

**Comment by gostaex** [comment]
It's clear. I tried like on version 11. Working. But it starts to work only after at least a complete unlocking of the device.

**Comment by Florlorez87** [comment]
On the last update The screen shot in fdroid, it shows something about days you can choose. I can’t find any of that in the app, and the tile doesn’t show also. What’s it I’m doing wrong?

**Comment by x13a** [comment]
> about days you can choose. I can’t find any of that

Make a long click on "Wipe on inactivity", you will get a chooser dialog.

> the tile doesn’t show

The tile now mimics to "Airplane mode". Do not forget to activate it in Triggers.

**Comment by x13a** [comment]
@InvisibleMan101 , @gostaex Duress Password is implemented as companion app here: [Duress](https://github.com/x13a/Duress). For now it is matching by password length, not by password itself. It is generic, to use it with Wasted you have to set action to: `me.lucky.wasted.action.TRIGGER`, receiver to: `me.lucky.wasted/CodeReceiver`, authentication code to the code from Wasted, password length to your actual password len plus at least two! Do not forget to activate Broadcast trigger in Wasted.

**Comment by Florlorez87** [comment]
Ah now i see the days options thank you it works…I find this one of the best options, in some countries it’s a must for e.g journalists etc. Just hope for them they won’t crack their password in 1 day before wasted can do the job.  

The airplane tile was there also I just couldn’t see it because it I didn’t click the edit icon, from there I see it is there.
Just trying to figure out  what it does because it is greyed out, when I click on it it does nothing it is greyed out, when I long press on it it opens the wasted Wasted App info.

**Comment by x13a** [comment]
> what it does because it is greyed out

You have to record a video of what are you doing and the Wasted app triggers screen.

**Comment by Florlorez87** [comment]
I was making a screen record to show you, clicked on airplane mode , turned black for a sec, then it said “erasing” lol. This is my second video so it’s working for sure.

I’m trying to figure it out so sorry for the amateur questions.
So I also got the panic shortcut as icon to the Home Screen when I click that it also wipes everything, what’s the difference with the airplane tile wipe?

after I’m familiar with that I’m gonna work a bit on the Broadcast option, just don’t want to ask To many questions at a time:)

https://user-images.githubusercontent.com/100235757/155309575-a6932e6b-ebfd-4c79-bf79-c20920f4a6ad.MOV


https://user-images.githubusercontent.com/100235757/155309646-fa296580-54af-4813-b7ed-9fd230574acf.MOV


https://user-images.githubusercontent.com/100235757/155309734-0294e189-18a2-40ea-841f-3b1e452ee277.MOV


https://user-images.githubusercontent.com/100235757/155309761-08fb0771-7f23-411c-8628-856da74fc4e6.MOV


https://user-images.githubusercontent.com/100235757/155309778-7d92fbe4-bf5c-4752-ac46-983972c035f5.MOV


https://user-images.githubusercontent.com/100235757/155309792-11242974-3aa9-4d53-adcc-4c2f2e437828.MOV


https://user-images.githubusercontent.com/100235757/155309807-3ff398c4-b98f-47d3-bd43-9174ca1013f8.MOV

**Comment by x13a** [comment]
The difference between `Tile` and `Shortcut` is that tiles are available when your device is locked too, while shortcuts are only available after you unlock it. Depending of your threat model you can choose what is right for you. `Tile` can be clicked by anyone who has phisical access to your device.

**Comment by x13a** [comment]
Read notifications permission, which you can find in `Settings > Notifications > Device and app notifications`, is required for a `Notification` trigger. `Broadcast` trigger is using broadcasts. This is an Android functionality to communicate between different apps on the same device.

**Comment by x13a** [comment]
> it says no such app

This links will be available after someone from F-Droid and Play Store will review and approve it. For now you have to use releases from GitHub: https://github.com/x13a/Duress/releases

**Comment by Florlorez87** [comment]
Ok done that also thanks very much!

To get it right: “ It will trigger on any password with the length you set. If your actual password is 4, than you set Duress to 6. When you write any 6 chars password it will trigger.”

so I set the password on screen lock I make for example: abcd, and when I enter the password 6 chars long at the Lock Screen to unlock it like : abcdef..then it must wipe/broadcast?

Cuz I’m trying and nothing happening.

**Comment by Florlorez87** [comment]
![1FF35919-BAFB-4A3A-808C-CE372CDD974B](https://user-images.githubusercontent.com/100235757/155730637-b450b49a-8168-4dfe-8520-eae6bbe40491.jpeg)
![6AA0F68D-760C-404A-BC6F-48C19782F585](https://user-images.githubusercontent.com/100235757/155730659-dc8b481e-1086-440d-9012-86daf80f9d3d.jpeg)

https://user-images.githubusercontent.com/100235757/155730768-77552218-0a77-4759-9fce-c355aea1b54e.MOV

**Comment by Florlorez87** [comment]
Uploading IMG_0023.MOV…

**Comment by Florlorez87** [comment]
https://user-images.githubusercontent.com/100235757/155730946-6c2304b2-6857-4172-b35d-272227f95de0.MOV

**Comment by Florlorez87** [comment]
Ok it did the job now, I was trying it in the work profile just in case, then I tried it in the main profile and got wiped everything :) amazing app, my respect!

**Comment by x13a** [comment]
> was trying it in the work profile

Thank you for letting me know. I will take a look what can be done to make it work in the work profile too.

**Comment by InvisibleMan101** [comment]
> Read notifications permission, which you can find in `Settings > Notifications > Device and app notifications`, is required for a `Notification` trigger. `Broadcast` trigger is using broadcasts. This is an Android functionality to communicate between different apps on the same device. The example is [Duress](https://github.com/x13a/Duress) app is using broadcast to trigger Wasted.

So just to triple check, if you send that authentication code from certain messaging apps with Duress setup with your settings and broadcast and notifications are on for Wasted this should wipe the device once it reads that code?

I tried this using briar messenger and couldn't get a code, is there certain messaging apps which this will and will not work with?

**Comment by x13a** [comment]
I haven't tested it with Briar, but Signal and Telegram worked for me.

**Comment by Florlorez87** [comment]
I’m lost with that, so how does that work.
I got signal/wasted/dures installed and activated on phone A. So do I send a message with signal app from phone B to Phone A with that authentication code and it wipes it?

**Comment by x13a** [comment]
You have to activate `Notification` trigger and gave Wasted access to notifications in `Settings > Notifications > Device and app notifications > Wasted`.

**Comment by Florlorez87** [comment]
It is activated, is it the way like I wrote before like sending a msg with signal? I tested it but didn’t check the wipe data box, maybe that’s the reason it didn’t do anything?

**Comment by Florlorez87** [comment]
![EE9CD286-8383-470D-8146-F674B3F9CEA6](https://user-images.githubusercontent.com/100235757/158342402-f42bcfad-9fdf-4a58-b86f-daeefe8385e0.jpeg)
![20F46FCF-DD61-4D6C-BCAC-851512F01DEC](https://user-images.githubusercontent.com/100235757/158342425-8730ddaf-7706-4a30-9971-3f590c9ffe19.jpeg)
![2F87FD35-3876-41D9-8F61-A44DCC2DE1AC](https://user-images.githubusercontent.com/100235757/158342434-2341cc5b-1d67-4b9e-be9d-13021ff9e96e.jpeg)

https://user-images.githubusercontent.com/100235757/158342444-0e444cf1-c3b5-43cf-9ad5-25336ab72b39.MOV


https://user-images.githubusercontent.com/100235757/158342452-1377d7b7-c513-460a-a3aa-f44d206a7f33.MOV

**Comment by x13a** [comment]
On your video this is not notifications permission I was talking about. Use search to find `Device and app notifications`.

https://img.webnots.com/2021/12/Notifictions-Menu.jpg

**Comment by Florlorez87** [comment]
This is so weird, I use lineage on a pixel 3a XL, android 11 and there is no such thing as privacy-Device and app notifications, even search doesn’t find it.
??

**Comment by x13a** [comment]
Download the app named Vibe. It has a bottom button, that will open this
menu for you.

**Comment by Florlorez87** [comment]
Yes now it worked, perfect. Thank you for all the help you really helped me with whenever I got stuck, thx.

**Comment by Florlorez87** [comment]
Question: the USB trigger which you are working on, what’s the idea?  Is it like that if any usb gets connected that it’s triggered: like let’s say a powerbank? 

Or is it like a configured usb?

**Comment by x13a** [comment]
I can not say for now how (and when) it will be implemented. Priority is to
resolve BFU for Sentry, Wasted and Duress. If any attached USB device has a
unique fingerprint, then it will be possible to implement some kind of a
whitelist. Or we can use the unlock device state as a bypass for the
trigger, but it may result in unexpected wiping. Without access to the
recent forensics hardware it is difficult to say what is the good way for
this. Because it can give a false feeling of safety. After the first unlock
Android stores encryption keys in memory. Is it enough to trigger on USB
attach or the data can be extracted without it? What data can be extracted
from the recent devices like Google Pixels and Samsung S series from a
live, locked (AFU) system? Does disabling USB data signaling is helping
here to mitigate risks? I am saying that there are many questions..

**Comment by Florlorez87** [comment]
Thank you for the clear explanation..I can’t wait until its finished .

“ After the first unlock
Android stores encryption keys in memory”

This is I guess why there is an auto reboot function in CalyxOS and graphene. It only doesn’t match up with the time in wasted. Wasted is like wipe if screen isn’t unlocked for 1 day .I set my device 1 day at wasted that’s why I don’t use auto reboot. It’s just an idea but maybe if wasted also can get the device wiped if not unlocked in x hours/minutes.
In some rogue state countries 1 day can be a long time is you were at a demonstration or if you are a journalist or working at the wrong company:) 
Just an idea..I’m very satisfied with the way it is now. Thx:)

**Comment by terra1122** [comment]
I've been trying for some time to use Duress with this to no avail. Wasted installed in work profile using shelter, have tried both Duress installed in normal profile and work.. no good. Android 12, CalyxOS. I assume Duress works with a text password only, no code

**Comment by Florlorez87** [comment]
Did you check this?
![19CF637F-AE09-4104-AC6E-70B2DE362ABA](https://user-images.githubusercontent.com/100235757/169713926-978066b7-d0d0-42fc-92c0-4b9aa1520d38.jpeg)

**Comment by terra1122** [comment]
Yep permissions are correct as far as I can tell with Duress. My 'Device & App' notifications are empty and I can't see a way to add things there.

**Comment by Florlorez87** [comment]
I’m not sure about it but I had the same issue before, I did copy paste the code, but duress didn’t work. Then I entered it one by one with keyboard, then it worked. Try different codes maybe that will fix it.

**Comment by terra1122** [comment]
> I’m not sure about it but I had the same issue before, I did copy paste the code, but duress didn’t work. Then I entered it one by one with keyboard, then it worked. Try different codes maybe that will fix it.

I'm not sure how you're able to copy and paste, I don't have that option although it would make things easier. I have tried many times different codes, all entered manually.

**Comment by terra1122** [comment]
> You have to activate `Notification` trigger and gave Wasted access to notifications in `Settings > Notifications > Device and app notifications > Wasted`.

In my Device and app notifications Wasted is not listed. How am I able to add it here?

**Comment by Florlorez87** [comment]
Yeah it don’t copy paste but I took a screenshot of the code and then highlight the code and copy it, but it never worked like that.
Mine works with shelter and work profile, just sometimes I didn’t so I deleted and started a new work profile/shelter..then it worked.

**Comment by x13a** [comment]
> So after further testing I can everything working perfectly, except when using work profiles and shelter. Is there a permission that i'm missing with Shelter?

Please explain what is not working for you?

**Comment by terra1122** [comment]
> > So after further testing I can everything working perfectly, except when using work profiles and shelter. Is there a permission that i'm missing with Shelter?
> 
> Please explain what is not working for you?

When using a work profile via shelter, I can't get Duress to trigger wasted. Works just fine without the work profile however.

**Comment by x13a** [comment]
I will take a look, thanks

**Comment by Florlorez87** [comment]
Thank you for the new updates! Especially the minutes and hours wipe when no activity function just made it perfect. 
and the broadcast code can be copied by longpress but I couldn’t seem to figure out how to generate a new code?

**Comment by x13a** [comment]
This functionality has been removed. Now for this you have to open app
settings in Android and clean it's data.

**Comment by x13a** [comment]
Implemented in [USBKill](https://github.com/x13a/USBKill) . Any discussion on USB trigger now there. Consider as alpha version.

**Comment by Florlorez87** [comment]
Thank you again, I will just wait until it’s downloadable as app, cant seem to download it like wasted or duress.

**Comment by x13a** [comment]
It is downloadable from https://github.com/x13a/USBKill/releases

**Comment by x13a** [comment]
Reopening, maybe will merge it here. Have not make a decision yet.

**Comment by xyz-nobody** [comment]
Test under GrapheneOS 

Thanks, it works perfectly unless you have Sentry.

As soon as I install Sentry on the smartphone, USBKIll does not work anymore. However the Broadcast is configured

**Comment by x13a** [comment]
Hmm, I am using it with Sentry..

**Comment by x13a** [comment]
Do you use Sentry option to disable USB data signaling?

**Comment by xyz-nobody** [comment]
I only use Sentry for the number of password attempt

But I can't disable the USB part because it conflicts with the grapheneos policy.
See ex

![sentry](https://user-images.githubusercontent.com/6586244/177383219-0ac3c126-9967-469f-a48a-2f25b07f611e.jpeg)

Sentry works perfectly.
But as soon as I add USBKill with Sentry it doesn't work anymore

**Comment by xyz-nobody** [comment]
I have already done a fresh install.
As soon as I install sentry, usbkill does not work anymore.

**Comment by x13a** [comment]
USBKill merged to Wasted. If you will find any bugs, please open a new issue. Closing

**Comment by Florlorez87** [comment]
Tried it on all 3 devices, “App not installed as package conflicts with an existing package”…uninstalled wasted, duress and usbkill, tried again but same warning.

**Comment by x13a** [comment]
What you have tried?

**Comment by Florlorez87** [comment]
Sorry forget to say what I tried. 
I tried the new download you posted 3 hours ago, with usbkill merged in wasted. 
it just doesn’t want to install.

**Comment by x13a** [comment]
Can not reproduce

**Comment by Florlorez87** [comment]
Thank you it’s working, new layout also perfect🥇

**Comment by florlorez** [comment]
Hi, I can’t seem to get around the new option Recast, do I fill in the same values like me.lucky.wasted.action.TRIGGER and so on..?
And I try to understand what it is, like instead of wipe send  a message..a message to where or what?  
I’m playing with it for a few days now but still no progress would appreciate it if I can get some help..thank you.

**Comment by x13a** [comment]
@florlorez Hi, _Recast_ option is used to send a **broadcast** message after a lock. All values are receiver dependant. A receiver can be any app that will listen for it. For example it can be Tasker, MacroDroid, Automate, KeyMapper. It is useful if you wanna to do some other actions instead of wipe. An app that will receive this broadcast has to do this actions.
